test/integration/volume/persistent_volumes_test.go

```diff
@@ -1488,31 +1467,40 @@ func waitForAnyPersistentVolumePhase(w watch.Interface, phase v1.PersistentVolum
 	}
 }
 
-func waitForAnyPersistentVolumeClaimPhase(w watch.Interface, phase v1.PersistentVolumeClaimPhase) {
-	for {
-		event := <-w.ResultChan()
-		claim, ok := event.Object.(*v1.PersistentVolumeClaim)
-		if !ok {
-			continue
-		}
-		if claim.Status.Phase == phase {
-			klog.V(2).Infof("claim %q is %s", claim.Name, phase)
-			break
+func waitForSomePersistentVolumeClaimPhase(ctx context.Context, testClient clientset.Interface, namespace string, objCount int, watchPVC watch.Interface, phase v1.PersistentVolumeClaimPhase) error {
+	return wait.ExponentialBackoffWithContext(ctx, retry.DefaultBackoff, func(ctx context.Context) (bool, error) {
+		for i := 0; i < objCount; i++ {
+			waitErr := waitForAnyPersistentVolumeClaimPhase(watchPVC, v1.ClaimBound)
+			if waitErr != nil {
+				klog.Errorf("Failed to wait for a claim (%d/%d) to be bound: %v", i+1, objCount, waitErr)
+				klog.Info("Recreating a watch for claims and re-checking the count of bound claims")
+				newWatchPVC, err := testClient.CoreV1().PersistentVolumeClaims(namespace).Watch(ctx, metav1.ListOptions{})
+				if err != nil {
+					return false, err
+				}
+				watchPVC.Stop()
+				watchPVC = newWatchPVC
+				return false, waitErr
```

Comment:
What is the purpose of this `if` branch? Why does it recreate the watch? Is this perhaps meant to continue polling after the apiserver has closed the watch (which can happen)?

Note that it doesn't achieve that: if `waitErr != nil`, then returning it in that last `return false, waitErr` causes `ExponentialBackoffWithContext` to return immediately with that error. I see that happening when race detection is enabled:

https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-integration-race-master/1975067824397225984

>     persistent_volumes_test.go:1101: Failed to wait for all claims to be bound: watch closed

It might work when turning that last return into `return false, nil`.

Reply:
See https://github.com/kubernetes/kubernetes/pull/134440